server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
 
     @Override
     protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
-        final long startTime = relativeTime();
-        final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
-
-        boolean hasIndexRequestsWithPipelines = false;
-        final MetaData metaData = clusterService.state().getMetaData();
-        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
-        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
-            IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
-            if (indexRequest != null) {
-                // get pipeline from request
-                String pipeline = indexRequest.getPipeline();
-                if (pipeline == null) {
-                    // start to look for default pipeline via settings found in the index meta data
-                    IndexMetaData indexMetaData = indicesMetaData.get(actionRequest.index());
-                    // check the alias for the index request (this is how normal index requests are modeled)
-                    if (indexMetaData == null && indexRequest.index() != null) {
-                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
-                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                            indexMetaData = alias.getWriteIndex();
-                        }
+        new BulkExecutor(task, bulkRequest, listener).run();
+    }
+
+    /**
+     * A runnable that will ensure the cluster state has been recovered enough to
+     * read index metadata and templates/pipelines.  Will retry up to the bulk's timeout
+     */
+    private final class BulkExecutor extends ActionRunnable<BulkResponse> {
+        private final ClusterStateObserver recoveredObserver;
+        private final BulkRequest bulkRequest;
+        private final Task task;
+        long startTime;
+
+        BulkExecutor(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+            super(listener);
+            this.recoveredObserver = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+            this.bulkRequest = bulkRequest;
+            this.task = task;
+            startTime = relativeTime();
+        }
+
+        @Override
+        protected void doRun() {
+            ClusterState currentState = clusterService.state();
```

Comment:
Technically I think we should get the state from `recoveredObserver.setAndGetObservedState()` to keep the state we observe in sync with the state that the `ClusterStateObserver` is observing. Alternatively, we could avoid constructing the `ClusterStateObserver` until we discover we need it, and give it the state we got from the cluster service - there's no need for it to be tracked in a field.